### PR TITLE
HPCC-17351 Logical file meta data migration tool

### PR DESCRIPTION
**Pull request by jakesmith**
Add option to daliadmin to map an existing group of
logical files onto a new cluster group.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

Tested with various small sets of files on small test clusters (on performance cluster nodes).
Needs further testing/feedback by ops.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@jakesmith Who can review this change?

**Comment by jakesmith** [comment]
@rwagner42 - please use/test.

@AttilaVamos - please review.

File: dali/daliadmin/daliadmin.cpp [hunk 1]
```diff
@@ -2808,6 +2809,240 @@ static void validateStore(bool fix, bool deleteFiles, bool verbose)
...
+        {
+            while (fileLists.ordinality()<=p)
+                fileLists.append(nullptr);
+
+            Linked<IFileIOStream> stream = fileLists.item(p);
+            if (nullptr == stream)
+            {
+                VStringBuffer filePartList("fileparts%u_%s_%u.lst", GetCurrentProcessId(), srcGroup.get(), p);
+                Owned<IFile> iFile = createIFile(filePartList);
+                Owned<IFileIO> iFileIO = iFile->open(IFOcreate);
```
Comment (AttilaVamos) [inline dali/daliadmin/daliadmin.cpp]: Should not this object creation be checked? If iFile is a valid pointer?

File: dali/daliadmin/daliadmin.cpp [hunk 2]
```diff
@@ -2808,6 +2809,240 @@ static void validateStore(bool fix, bool deleteFiles, bool verbose)
...
+
+                            Owned<IFileIOStream> iFileIOStream = getFileIOStream(relPos);
+
+                            StringBuffer outputLine;
+                            srcEp.getIpText(outputLine);
+                            outputLine.append(",");
+                            outputLine.append(dstIpStr);
+                            outputLine.append(",");
+
+                            IPartDescriptor *part = fileDesc->queryPart(p);
```
Comment (AttilaVamos) [inline dali/daliadmin/daliadmin.cpp]: Can '*part' be nullptr?

**Review by AttilaVamos — COMMENTED** [review]
There are several questions.

**Comment by AttilaVamos** [comment]
@jake @back to you with some questions.

File: dali/daliadmin/daliadmin.cpp [hunk 3]
```diff
@@ -129,6 +129,7 @@ void usage(const char *exe)
   printf("  wuidcompress <wildcard> <type>  --  scan workunits that match <wildcard> and compress resources of <type>\n");
   printf("  wuiddecompress <wildcard> <type> --  scan workunits that match <wildcard> and decompress resources of <type>\n");
   printf("  xmlsize <filename> [<percentage>] --  analyse size usage in xml file, display individual items above 'percentage' \n");
+  printf("  migratefiles <src-group> <target-group> [<filemask>] [DRYRUN] [CREATEMAPS] [LISTONLY] [VERBOSE]\n");
```
Comment (jakesmith) [inline dali/daliadmin/daliadmin.cpp]: could change usage print, but tests are case-insensitive, so no need for conversion.

File: dali/daliadmin/daliadmin.cpp [hunk 4]
```diff
@@ -2808,6 +2809,240 @@ static void validateStore(bool fix, bool deleteFiles, bool verbose)
...
+
+    enum class mg_options : unsigned { nop, createmaps=1, listonly=2, dryrun=4, verbose=8};
+
+    StringArray options;
+    options.appendList(_options, ",");
+    mg_options opts = mg_options::nop;
+    ForEachItemIn(o, options)
+    {
+        const char *opt = options.item(o);
+        if (strieq("CREATEMAPS", opt))
```
Comment (jakesmith) [inline dali/daliadmin/daliadmin.cpp]: strieq is case insensitive test.

File: dali/daliadmin/daliadmin.cpp (hunk 1 above)
Comment (jakesmith) [inline dali/daliadmin/daliadmin.cpp]: iFile can't be null, either valid or exception.
iFileIO could be I think.
I'll add a catch/throw exception here.

File: dali/daliadmin/daliadmin.cpp [hunk 5]
```diff
@@ -2808,6 +2809,240 @@ static void validateStore(bool fix, bool deleteFiles, bool verbose)
...
+            Owned<IFileDescriptor> fileDesc = deserializeFileDescriptorTree(&root, &queryNamedGroupStore());
+            unsigned numClusters = fileDesc->numClusters();
+            for (unsigned g=0; g<numClusters; g++)
+            {
+                StringBuffer clusterGroupName;
+                fileDesc->getClusterGroupName(g, clusterGroupName);
+                const char *sub = strchr(clusterGroupName, '[');
+                StringAttr group;
+                if (sub)
+                    group.set(clusterGroupName, sub-clusterGroupName);
```
Comment (jakesmith) [inline dali/daliadmin/daliadmin.cpp]: this is common style (no spaces) elsewhere

File: dali/daliadmin/daliadmin.cpp [hunk 6]
```diff
@@ -2808,6 +2809,240 @@ static void validateStore(bool fix, bool deleteFiles, bool verbose)
...
+                        if (cluster)
+                            cluster->setProp("@name", _tgtGroup);
+                        else
+                            WARNLOG("No Cluster found for file: %s", name.str());
+                    }
+                    unsigned parts = fileDesc->numParts();
+                    unsigned dstPos = 0;
+                    for (unsigned p=0; p<parts; p++)
+                    {
+                        const SocketEndpoint &srcEp = fileDesc->queryPart(p)->queryNode()->endpoint();
```
Comment (jakesmith) [inline dali/daliadmin/daliadmin.cpp]: this is deserialized file descriptor neither part or node can be null here.

File: dali/daliadmin/daliadmin.cpp (hunk 2 above)
Comment (jakesmith) [inline dali/daliadmin/daliadmin.cpp]: no

File: dali/daliadmin/daliadmin.cpp [hunk 7]
```diff
@@ -2808,6 +2809,240 @@ static void validateStore(bool fix, bool deleteFiles, bool verbose)
...
+            if (mgOpt(mg_options::listonly))
+                return;
+
+            bool doCommit = false;
+            StringBuffer _tgtClusterGroupText = tgtClusterGroupText;
+            Linked<IGroup> _tgtClusterGroup = tgtClusterGroup;
+
+            Owned<IFileDescriptor> fileDesc = deserializeFileDescriptorTree(&root, &queryNamedGroupStore());
+            unsigned numClusters = fileDesc->numClusters();
+            for (unsigned g=0; g<numClusters; g++)
```
Comment (jakesmith) [inline dali/daliadmin/daliadmin.cpp]: I'll change.

**Comment by jakesmith** [comment]
@AttilaVamos - could you take a look/review the new commit. Which are mostly changes following testing against Dataland DFS.

File: dali/daliadmin/daliadmin.cpp (hunk 4 above)
Comment (AttilaVamos) [inline dali/daliadmin/daliadmin.cpp]: Right

File: dali/daliadmin/daliadmin.cpp (hunk 1 above)
Comment (AttilaVamos) [inline dali/daliadmin/daliadmin.cpp]: Ok, please add catch/throw exception.

**Review by AttilaVamos — CHANGES_REQUESTED** [review]
One changes remained.

**Comment by jakesmith** [comment]
@AttilaVamos - added check for failing to open file.
Other changes are just moving code that was only used in inner block and minor change to logging.

Please check and I'll squash ready for merge.

**Comment by AttilaVamos** [comment]
@jakesmith it seems good. Do you plan to add more commit?

**Comment by jakesmith** [comment]
No, I will squash it in a moment.

**Comment by jakesmith** [comment]
@richardkchapman - please merge.

**Pull request merged**